File: admin/src/api/youtube.ts
import { api } from "./client";

export interface TrackProgress {
  index: number;
  title: string;
  status: string;
  error?: string;
  content_id?: number;
}

export interface DownloadTask {
  task_id: string;
  url: string;
  status: string;
  tracks: TrackProgress[];
  completed_count: number;
  failed_count: number;
  total_count: number;
  error?: string;
  created_at: number;
}

export interface YouTubeDownloadParams {
  url: string;
  content_type: string;
  category_id: number;
  artist_name?: string;
  artist_type: string;
  tag_ids?: number[];
  age_min?: number;
  age_max?: number;
}

// ========== 搜索相关类型 ==========

export interface SearchParams {
  keyword: string;
  platforms?: string[];
  content_type?: string;
  max_results?: number;
}

export interface SearchResultItem {
  platform: string;
  url: string;
  title: string;
  duration: number;
  view_count: number;
  like_count: number;
  thumbnail?: string;
  uploader?: string;
  upload_date?: string;
  quality_score: number;
  exists_in_db: boolean;
}

export interface SearchResponse {
  results: SearchResultItem[];
  total_count: number;
  platforms_searched: string[];
  dedup_removed_count: number;
}

export interface BatchDownloadParams {
  urls: string[];
  content_type: string;
  category_id: number;
  artist_name?: string;
  artist_type: string;
  tag_ids?: number[];
  age_min?: number;
  age_max?: number;
}

export interface PlatformInfo {
  id: string;
  label: string;
}

// ========== 原有 API (保持兼容) ==========

export const youtubeApi = {
  startDownload: async (params: YouTubeDownloadParams): Promise<DownloadTask> => {
    const response = await api.post("/api/v1/admin/youtube/download", params);
    return response.data;
  },

  getTask: async (taskId: string): Promise<DownloadTask> => {
    const response = await api.get(`/api/v1/admin/youtube/tasks/${taskId}`);
    return response.data;
  },

  listTasks: async (): Promise<DownloadTask[]> => {
    const response = await api.get("/api/v1/admin/youtube/tasks");
    return response.data;
  },

  cancelTask: async (taskId: string): Promise<void> => {
    await api.post(`/api/v1/admin/youtube/tasks/${taskId}/cancel`);
  },
};

// ========== 新增 API：搜索 + 批量下载 ==========

export const downloadApi = {
  search: async (params: SearchParams): Promise<SearchResponse> => {
    const response = await api.post("/api/v1/admin/download/search", params);
    return response.data;
  },

  batchDownload: async (params: BatchDownloadParams): Promise<DownloadTask> => {
    const response = await api.post("/api/v1/admin/download/batch", params);
    return response.data;
  },

  getPlatforms: async (): Promise<PlatformInfo[]> => {
    const response = await api.get("/api/v1/admin/download/platforms");
    return response.data;
  },
};

File: server/app/api/routes/youtube.py
"""
内容下载 API 路由
"""

import logging

from fastapi import APIRouter, Depends

from ...models.schemas import YouTubeDownloadRequest, SearchRequest, BatchDownloadRequest
from ...models.response import ErrorCode, BusinessException, success_response
from ..deps import get_download_service
from ...services.download_service import DownloadService

router = APIRouter()
logger = logging.getLogger(__name__)


# ========== 原有端点 (保持兼容) ==========

@router.post("/api/v1/admin/youtube/download")
async def start_download(
    request: YouTubeDownloadRequest,
    download_service: DownloadService = Depends(get_download_service),
):
    """创建下载任务（URL 下载）"""
    task = download_service.start_download(
        url=request.url,
        content_type=request.content_type,
        category_id=request.category_id,
        artist_name=request.artist_name,
        artist_type=request.artist_type,
        tag_ids=request.tag_ids,
        age_min=request.age_min,
        age_max=request.age_max,
    )
    return success_response(data=task.to_dict())


@router.get("/api/v1/admin/youtube/tasks")
async def list_tasks(
    download_service: DownloadService = Depends(get_download_service),
):
    """获取所有下载任务"""
    tasks = download_service.list_tasks()
    return success_response(data=tasks)


@router.get("/api/v1/admin/youtube/tasks/{task_id}")
async def get_task(
    task_id: str,
    download_service: DownloadService = Depends(get_download_service),
):
    """获取单个任务进度"""
    task = download_service.get_task(task_id)
    if not task:
        raise BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "Task not found")
    return success_response(data=task.to_dict())


@router.post("/api/v1/admin/youtube/tasks/{task_id}/cancel")
async def cancel_task(
    task_id: str,
    download_service: DownloadService = Depends(get_download_service),
):
    """取消下载任务"""
    success = download_service.cancel_task(task_id)
    if not success:
        raise BusinessException(ErrorCode.INVALID_PARAMS, "无法取消该任务")
    return success_response(message="Task cancelled")


# ========== 新增端点：搜索 + 批量下载 ==========

@router.post("/api/v1/admin/download/search")
async def search_content(
    request: SearchRequest,
    download_service: DownloadService = Depends(get_download_service),
):
    """多平台关键字搜索"""
    result = await download_service.search(
        keyword=request.keyword,
        platforms=request.platforms,
        content_type=request.content_type,
        max_results=request.max_results,
    )
    return success_response(data=result)


@router.post("/api/v1/admin/download/batch")
async def batch_download(
    request: BatchDownloadRequest,
    download_service: DownloadService = Depends(get_download_service),
):
    """批量下载选中项"""
    if not request.urls:
        raise BusinessException(ErrorCode.INVALID_PARAMS, "至少选择一个内容")

    task = download_service.start_batch_download(
        urls=request.urls,
        content_type=request.content_type,
        category_id=request.category_id,
        artist_name=request.artist_name,
        artist_type=request.artist_type,
        tag_ids=request.tag_ids,
        age_min=request.age_min,
        age_max=request.age_max,
    )
    return success_response(data=task.to_dict())


@router.get("/api/v1/admin/download/platforms")
async def get_platforms(
    download_service: DownloadService = Depends(get_download_service),
):
    """获取可用搜索平台列表"""
    platforms = download_service.get_available_platforms()
    return success_response(data=platforms)
